packages/xl-ai/src/prosemirror/agent.ts

```diff
@@ -0,0 +1,325 @@
+      throw new Error(
+        "Slice has openStart or openEnd > 0, but structure=false",
+      );
+    }
+
+    // 1. Select text to be removed/replaced
+    agentSteps.push({
+      prosemirrorSteps: [],
+      selection: {
+        anchor: tr.mapping.map(step.from),
+        head: tr.mapping.map(step.to),
+      },
+      type: "select",
+    });
+
+    // Replace the content
+    const sliceTextContent = step.slice.content.textBetween(0, step.slice.size);
+
+    const alreadyHasSameText =
+      sliceTextContent ===
+      tr.doc.textBetween(tr.mapping.map(step.from), tr.mapping.map(step.to));
+
+    let sliceTo: number;
+
+    if (alreadyHasSameText) {
+      sliceTo = step.slice.content.size; // replace all at once, it's probably a mark update
+    } else if (sliceTextContent.length === 0) {
+      sliceTo = step.slice.content.size; // there's no replacement text, so use entire slice as replacement
+    } else {
+      // replace with the first character (similar to how a user would do it when selecting text and starting to type)
+      const firstCharIndex = getFirstChar(step.slice.content);
+      if (firstCharIndex === undefined) {
+        // should have been caught by previous if statement
+        throw new Error("unexpected: no first character found");
+      }
+      sliceTo = firstCharIndex + 1;
+    }
+
+    let replaceEnd = tr.mapping.map(step.to);
+    const replaceFrom = tr.mapping.map(step.to);
+    let first = true;
+
+    // Note that below, we don't actually delete content, but we mark it as deleted
+    // with a deletion mark. Similarly, we mark inserted content with an insertion mark.
+    // It might be cleaner to;
+    // a) make this optional
+    // b) actually delete / insert the content and let prosemirror-suggest-changes handle the marks
+    for (let i = sliceTo; i <= step.slice.content.size; i++) {
+      // in the first agent step (first character replacement) we mark existing content (if any) as deleted
+      // (step.from !== step.to is used to check for replacements vs inserts)
+      const isReplacing = first && step.from !== step.to;
+
+      const stepIndex = tr.steps.length;
+      if (isReplacing) {
+        const $pos = tr.doc.resolve(tr.mapping.map(step.from));
+        if ($pos.nodeAfter?.isBlock) {
+          // mark the entire node as deleted. This can be needed for inline nodes or table cells
+          tr.addNodeMark($pos.pos, pmSchema.mark("deletion", {}));
+        }
+        tr.addMark($pos.pos, replaceEnd, pmSchema.mark("deletion", {}));
+        replaceEnd = tr.mapping.map(step.to);
+      }
+
+      // note, instead of inserting one charachter at a time at the end (a, b, c)
+      // we're replacing the entire part every time (a, ab, abc)
+      // would be cleaner to do just insertions, but didn't get this to work with the add operation
+      // (and this kept code relatively simple)
+      const replacement = new Slice(step.slice.content.cut(0, i), 0, 0);
```

Comment:
This will affect how the ops are laid out in the undo/redo stack.
Here is where in prosemirror-history steps are merged together: https://github.com/ProseMirror/prosemirror-history/blob/master/src/history.ts#L237
This is a ReplaceStep's merge method: https://github.com/ProseMirror/prosemirror-transform/blob/137ff74738bd1b50d49416cd6cfdbbf52cb059ef/src/replace_step.ts#L48-L62
So, this can only merge ops which occur one after the other, not overlapping ranges like you've got here.

Reply:
Let's discuss undo / redo tomorrow morning - let's look at the zoomed out issue first because I think the way undo / redo works in prosemirror might not be applicable to our use-case at all